component/http/handler.go

```diff
@@ -94,7 +96,7 @@ func determineEncoding(h http.Header) (string, encoding.DecodeFunc, encoding.Enc
 		}
 		// No valid headers found after going through all headers
 		if !encFound {
-			return "", nil, nil, errors.New("accept header not supported")
+			return "", nil, nil, ErrAcceptHeaderNotSupported
```

Comment:
What's the rationale for using a sentinel error here instead of `errors.New`?

Reply:
If it's just for handiness regarding testing, I'm really not convinced about the rationale for creating an exported sentinel error.
Also, string comparison is involved when comparing sentinel errors:

```

var (
	ErrFoo = errors.New("accept header not supported")
	ErrBar = errors.New("accept header not supported")
)	

func TestEq(t *testing.T) {
	assert.Equal(t, ErrFoo, ErrBar)
}
```

This test is valid.

Again, not convinced. If we don't need external clients to check for a specific error value, I don't think we need to create a custom error value. Otherwise, could be an open door for always creating error values.

Reply:
As I see it creating error values has de following advantages:
1. consistency, the error is exactly the same whenever is used
2. ease of maintenance, the error can be used in multiple places, but only defined in one, so if we need to change it in the future it only has to be changed once
3. assertment, `asset.Equal` might use string comparison but `errors.(Is|As)` do not

**If we don't want to expose it as part of our api I can rename it as `camelCase` so it doesn't get exported**

Can you see any drawback about using predefined errors above define them inline?

Reply:
As for (1), it also has a little performance improvement has memory is not allocated when the same error object (reference) is re-used.

Can this error ever reach the caller (with or without wrapping) or is it purely internal?

Reply:
I do see a drawback, forcing us to create an error value whenever this error value has to be tested.

We can take examples from the stdlib:
* Error returned: https://github.com/golang/go/blob/2580d0e08d5e9f979b943758d3c49877fb2324cb/src/crypto/x509/verify.go#L80
* Error tested: https://github.com/golang/go/blob/feb024f4153395e5bbb2a51bb3d1ddc4f5b0d2dc/src/crypto/x509/root_darwin_test.go#L87

They don't create an error value for every single error they want to check.

Reply:
You are right, they are just using the string value for identifying to error
Previously we had the same errors defined in 2 different places:
- https://github.com/GerardRodes/patron/blob/master/component/http/handler.go#L97
- https://github.com/GerardRodes/patron/blob/master/component/http/handler.go#L119

It makes me think that that error should be defined in a single point or it could deviate into different alternatives of the same error message, ie:
- "accept header not supported"
- "invalid accept header"

In this case it won't, probably, since they are close one from other, but I see it as a good practice to only declare the same error once.

---

I appreciate that stdlib reference, it's cool to learn from there

Reply:
> or it could deviate into different alternatives of the same error message

If they deviate, we can catch it in unit tests.

> but I see it as a good practice to only declare the same error once.

I disagree, for the reasons I mentioned. Otherwise, we will always create a custom error value every time we return a specific error that we want to check in unit tests and that will be cumbersome. I don't think it's the philosophy of error handling in Go.

Reply:
:+1: I switch back to inlined errors